[mono] inflate function pointer types; hang indirect call wrappers on the <Module>

lambdageek opened the pull request:
1. We were missing a case for `MONO_TYPE_FNPTR` in `inflate_generic_type`

2. When an indirect call was inside of a method that was part of a generic class, we were creating wrapper methods for the indirect calls parented by the GTD type of the caller.  This ended up making common_call_trampoline consider the wrapper as needing an mrgctx.  But if the caller was actually a normal closed instance type no mrgctx is set up.  So we end up dereferencing a null pointer.  Instead, when we create an indirect call wrapper method (which we asserted has no type params and which is always just sets up the GC transition and forwards the arguments to the indirect callee - so it never depends on the generic context), make its parent be the <Module> class of the calling assembly.  This is consistent with how we cache these indirect methods (in the calling image, by the signature of the indirect callee) - it has nothing to do with the caller class.

Fixes https://github.com/dotnet/runtime/issues/106811

lambdageek committed:
[mono] inflate function pointer types; hang indirect call wrappers on the <Module>

1. We were missing a case for `MONO_TYPE_FNPTR` in `inflate_generic_type`

2. When an indirect call was inside of a method that was part of a
generic class, we were creating wrapper methods for the indirect calls
parented by the GTD type of the caller.  This ended up making
common_call_trampoline consider the wrapper as needing an mrgctx.  But
if the caller was actually a normal closed instance type no mrgctx is
set up.  So we end up dereferencing a null pointer.  Instead, when we
create an indirect call wrapper method (which we asserted has no type
params and which is always just sets up the GC transition and forwards
the arguments to the indirect callee - so it never depends on the
generic context), make its parent be the <Module> class of the calling
assembly.  This is consistent with how we cache these indirect
methods (in the calling image, by the signature of the indirect
callee) - it has nothing to do with the caller class.

Fixes https://github.com/dotnet/runtime/issues/106811

lambdageek commented:
Need a testcase, I think.

Here's a self-contained repro:

```csharp
 System;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;


namespace HelloWorld
{
    public static class InteropUtil
    {
        [UnmanagedCallersOnly]
        private static void Bar(int x) => Console.WriteLine ($"Bar {x}");
        
        //[DllImport("libnative.dylib")]
        public unsafe static /*extern*/ IntPtr getNativeFunctionPointer() => (IntPtr)(void*)(delegate* unmanaged<int,void>)&InteropUtil.Bar;
    }

    unsafe public class NativeInvoker<T>
    {
        [MethodImpl(MethodImplOptions.NoInlining)]
        T Foo(T x) => x;
        public void Invoke(IntPtr fptr, T arg)
        {
            T x = Foo(arg);
            var func = (delegate* unmanaged<T, void>)fptr;
            func(x);
        }
    }
    internal class Program
    {
        private static void Main(string[] args)
        {
            var invoker = new NativeInvoker<int>();
            invoker.Invoke(InteropUtil.getNativeFunctionPointer(), 42);
        }
    }
}
```

kg commented on src/mono/mono/metadata/class.c:
```diff
@@ -870,6 +870,21 @@ inflate_generic_type (MonoImage *image, MonoType *type, MonoGenericContext *cont
+		MonoMethodSignature *new_sig = mono_inflate_generic_signature (in_sig, context, error);
+		if (!new_sig || !is_ok (error))
+			return NULL;
+		if (new_sig == in_sig)
```
can this happen? i'm not opposed to it but it seems weird

lambdageek replied:
Yes.
1. The signature inflation code has a cache, so repeated values are possible, in general.
2. The incoming context could put in the same type variables that are already in the sig, so the result is the same as the input.

kg reviewed (approved):
Code looks correct, not certain about the nuances though

lambdageek commented on src/mono/mono/metadata/class.c:
```diff
@@ -870,6 +870,21 @@ inflate_generic_type (MonoImage *image, MonoType *type, MonoGenericContext *cont
+		MonoMethodSignature *in_sig = type->data.method;
+		// quick bail out - if there are no type variables anywhere in the signature,
+		// there's nothing that could get inflated.
+		if (!in_sig->has_type_parameters)
+			return NULL;
+		MonoMethodSignature *new_sig = mono_inflate_generic_signature (in_sig, context, error);
+		if (!new_sig || !is_ok (error))
+			return NULL;
+		if (new_sig == in_sig)
+			return type;
+		MonoType *nt = mono_metadata_type_dup (image, type);
+		nt->data.method = new_sig;
+		return nt;
+	}
```
all of the returns here need to be more careful: if `changed` is true, that means the cmods on the `MonoType` got inflated so we have to return either `type` or `nt`.  if `!changed` then we should return `NULL` or `nt`

jkurdek replied:
We should be good reusing the control flow of `MONO_TYPE_PTR` case, right?

jkurdek reviewed (approved):
Thanks for looking into that!

lambdageek committed:
pay attention to `changed` when inflating

if the cmods changed, but the sig didn't we want to return `type`.
return NULL only if there's an error or if nothing else changed

lambdageek committed:
add regression test

srxqds commented:
should this consider merge to release/9.0?

lambdageek commented:
/backport to release/9.0

lambdageek commented:
/backport to release/8.0-staging
